build/source/engine/systemSolv.f90

```diff
@@ -339,292 +383,529 @@ real(dp)                        :: fOld,fNew                    ! function value
+
+  ! make association with model indices defined in indexSplit
+  stateSubset: associate(&
+  ixSnowSoilNrg      => indx_data%var(iLookINDEX%ixSnowSoilNrg)%dat,     & ! intent(in): [i4b(:)] indices for energy states in the snow+soil subdomain
+  ixSnowOnlyHyd      => indx_data%var(iLookINDEX%ixSnowOnlyHyd)%dat,     & ! intent(in): [i4b(:)] indices for hydrology states in the snow subdomain
+  ixStateType_subset => indx_data%var(iLookINDEX%ixStateType_subset)%dat & ! intent(in): [i4b(:)] [state subset] type of desired model state variables
+  ) ! associations
+  
+  ! allocate space for solution and scaling vectors
+  allocate(stateVecInit(nSubset), stateVecTrial(nSubset), stateVecNew(nSubset), fluxVec0(nSubset), fluxVecNew(nSubset), fScale(nSubset), xScale(nSubset), stat=err)
+  if(err/=0)then; err=20; message=trim(message)//'unable to allocate space for the solution and scaling vectors'; return; endif
+
+  ! allocate space for the diagonal matrix, multipliers and residual vectors
+  allocate(dMat(nSubset), sMul(nSubset), rVec(nSubset), rAdd(nSubset), resSinkNew(nSubset), resVecNew(nSubset), stat=err)
+  if(err/=0)then; err=20; message=trim(message)//'unable to allocate space for the diagonal matrix, multipliers, and residual vectors'; return; endif
+
+  ! identify the matrix solution method
+  ! NOTE: uses nSubset so must be defined here
+  ! (the type of matrix used to solve the linear system A.X=B)
+  if(local_ixGroundwater==qbaseTopmodel .or. testBandDiagonal .or. forceFullMatrix)then
+   nLeadDim=nSubset        ! length of the leading dimension
+   ixMatrix=ixFullMatrix   ! named variable to denote the full Jacobian matrix
+  else
+   nLeadDim=nBands         ! length of the leading dimension
+   ixMatrix=ixBandMatrix   ! named variable to denote the band-diagonal matrix
+  endif
+
+  ! define the mask of the fluxes used
+  do iVar=1,size(flux_meta)
+   fluxMask(iVar) = any(ixStateType_subset==flux2state_meta(iVar)%state1) .or. any(ixStateType_subset==flux2state_meta(iVar)%state2)
+  end do
+
+  ! initialize the model fluxes (some model fluxes are not computed in the iterations)
+  do iVar=1,size(flux_meta)
+   ! copy over the master flux structure since some model fluxes are not computed in the iterations
+   flux_temp%var(iVar)%dat(:) = flux_data%var(iVar)%dat(:)
+   ! set master flux vector to zero for the fluxes computed here
+   if(fluxMask(iVar)) flux_data%var(iVar)%dat(:) = 0._dp
+  end do
+
+  ! -----
+  ! * variable-dependent sub-stepping...
+  ! ------------------------------------
+  
+  ! define the number of substeps for each solution
+  if(ixSplitOption==deCoupled_nrgMass)then
+   select case(iSplit)
+    case(nrgSplit);  nSubstep=1
+    case(massSplit); nSubstep=1
+    case default; err=20; message=trim(message)//'unable to identify splitting operation'; return
+   end select
+  endif
+
+  ! define the length of the substep
+  dtSplit = dt/real(nSubstep, kind(dp))
+
+  ! loop through substeps
+  do iSubstep=1,nSubstep
```

Comment:
Hi @michaelou — the intent is to have variable-dependent sub-stepping at this point. The dynamic substepping for the model as a whole is handled in `coupled_em.f90`